invisible_cities/detector_simulation/buffer_functions.py

```diff
@@ -0,0 +1,149 @@
+import numpy  as np
+import pandas as pd
+
+from typing    import  Callable
+from typing    import Generator
+from typing    import      List
+from typing    import     Tuple
+
+from .. core                import                system_of_units as units
+from .. reco.peak_functions import indices_and_wf_above_threshold
+from .. reco.peak_functions import                 split_in_peaks
+
+
+def weighted_histogram(data: pd.DataFrame, bins: np.ndarray) -> np.ndarray:
+    return np.histogram(data.time, weights=data.charge, bins=bins)[0]
+
+
+def bin_sensors(sensors   : pd.Series,
+                bin_width : float    ,
+                t_min     : float    ,
+                t_max     : float    ,
+                max_buffer: int      ) -> Tuple:
+    """
+    Raw data binning function.
+
+    Parameters
+    ----------
+    sensors : List of Waveforms for one sensor type
+    Should be sorted into one type/binning
+    t_min   : float
+              Minimum time to be used to define bins.
+    t_max   : float
+              As t_min but the maximum to be used
+    """
+    max_time    = min(t_max, t_min + max_buffer)
+    min_bin     = np.floor(t_min    / bin_width) * bin_width
+    max_bin     = np.ceil (max_time / bin_width) * bin_width
+
+    bins        = np.arange(min_bin, max_bin, bin_width)
+    b_pad       = bins[-1] + bin_width
+    bin_sensors = sensors.groupby('sensor_id').apply(weighted_histogram    ,
+                                                     np.append(bins, b_pad))
+    return bins, bin_sensors
```

Comment:
Maybe there is an edge case where this doesn't work (?) but, isn't it simpler:
```suggestion
    bins        = np.arange(min_bin, max_bin + bin_width, bin_width)
    bin_sensors = sensors.groupby('sensor_id').apply(weighted_histogram, bins)
    return bins[:-1], bin_sensors
```

Also, is it intended to not return the last upper bin edge?